Spec/Push.swift

```diff
@@ -76,6 +79,70 @@ class Push : QuickSpec {
                 }
             }
 
+            // RSH2d
+            it("should occur a GettingPushDeviceDetailsFailed event when the APNS registration fails") {
+                let stateMachine = rest.push.activationMachine()
+                let stateMachineDelegate = StateMachineDelegate()
+                stateMachine.delegate = stateMachineDelegate
+                defer {
+                    stateMachine.transitions = nil
+                    stateMachine.delegate = nil
+                }
+
+                let expectedError = ARTErrorInfo(domain: ARTAblyErrorDomain, code: 1234, userInfo: nil)
+
+                let hook = stateMachine.testSuite_injectIntoMethod(after: NSSelectorFromString("registerForRemoteNotifications")) {
+                    AblyTests.extraQueue.async(execute: {
+                        ARTPush.didFailToRegisterForRemoteNotificationsWithError(expectedError, rest: rest)
+                    })
+                }
+                defer { hook.remove() }
+
+                waitUntil(timeout: testTimeout) { done in
+                    stateMachine.transitions = { event, _, _ in
+                        if let event = event as? ARTPushActivationEventGettingPushDeviceDetailsFailed {
+                            expect(event.error) == expectedError
+                            done()
+                        }
+                    }
+                    rest.push.activate()
+                }
+            }
+
+            // RSH2e
+            it("should verify the validity of the push device details") {
```

Comment:
@paddybyers I noticed that this isn't well implemented. The spec says **_Each time the library is instanced_** and I was doing the check when `activate` is called. So I did that change and it's not recommended because the instance needs to  access the `device` property in the initialiser. Some of the `device` properties (like the device secret) will be loaded from disk, more precisely from the iOS Keychain and that could be a problem. Since we can create an instance right before the app is running, the Keychain is not ready to be used and errors like `ERROR: Device Secret couldn't be updated (SecBase.h I/O error)` can happen. Maybe the best would be adding a `[rest/realtime].push.setup()` or `[rest/realtime].push.configure()` that should be called in the AppDelegate like other libraries do.

I.e. Firebase:

![Screenshot 2019-08-07 at 11 30 32](https://user-images.githubusercontent.com/3541185/62616197-dec67080-b906-11e9-9dfa-2d5db0ae0bb5.png)

Reply:
@ricardopereira But wasn't `activate` the equivalent of that `configure` method you propose?

Reply:
@tcard Not quite because the user can call `activate()` whatever he/she wants and the `configure()` method was meant only for the initialisation of the app.

> paddy 2:11 PM
> There will be a further push spec change where we re-do the registration request each time activate() is called, even if already registered

Even so, this spec will change so this is in stand by.